### PR TITLE
be compatible with FQCN services in Symfony 3.3

### PR DESCRIPTION
**Pull request by xabbuh**
This *could*  be an alternative to symfony/symfony#23239.

### CODE REVIEW

**Comment by stof** [comment]
if you want to use the `::` notation, meaning **not** using the controller as a service, the right solution would be to change the order of conditions in the if/elseif chain instead. This will also avoid getting the service from the container to get the class.

**Comment by xabbuh** [comment]
But the goal is to use the controller as a service. This is just a workaround around some inconsistency in FrameworkBundle's `ControllerResolver` implementation. And I am not really convinced that this is the right fix. Probably it's just wrong that `setContainer()` is called for a controller being registered as a service, but I am not sure we should revert that now.

**Comment by stof** [comment]
@xabbuh but your code now is equivalent to witching the order, except for the case where you have a service id equal to a class name, but where the actual service class is not equal to the id (in which case, I'm not sure your code behaves the expected way either)

**Comment by xabbuh** [comment]
We should probably only do this when the id equals the class.

File: Routing/Loader/RestRouteLoader.php [hunk 1]
```diff
@@ -126,6 +126,10 @@ private function getControllerLocator($controller)
             if ($useScope) {
                 $this->container->leaveScope('request');
             }
+
+            if ($controller === $class && class_exists($controller)) {
```
Comment (GuilhemN) [inline Routing/Loader/RestRouteLoader.php]: I don't think we need the class existence check as $class comes from php itself.

Comment (xabbuh) [inline Routing/Loader/RestRouteLoader.php]: indeed

**Comment by GuilhemN** [comment]
It seems the tests need to be updated :)

**Comment by xabbuh** [comment]
It seems tests are passing now. :)

**Comment by xabbuh** [comment]
closing as symfony/symfony#23239 has been merged

**Pull request closed without merging**